awslabs/aws-lambda-powertools-typescript
# feat(logger): make loglevel types stricter

## Description of your changes

The Logger constructor accepts an optional `logLevel`  parameter with [default](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/packages/logger/src/Logger.ts#L120) to `INFO`. While the library already has types for all the log levels, and performs a validation of the value passed to `logLevel`, the constructor currently accepts any type of value.

As detailed in #1309, this can lead to issues that go undetected at edit/build time since TypeScript is not being leveraged to validate these inputs. For instance, currently it's possible to pass the following value and Logger will accept the value and silently default to `INFO`:
```ts
const logger = new Logger({ logLevel: 'DEBG' }); // instead of DEBUG
```

This PR, that once merged will close #1309, narrows the types for `logLevel` so that only certain values both upper and lower case variations are accepted. If an invalid value is passed, users will see errors both in their IDE and at compile time.

For the time being I have decided to not include a warning log in case of invalid value, given that we are already narrowing the types. However I'm open to reconsider this decision in a future PR if there's demand.

### How to verify this change

See checks under this PR, as well as the [results of the integration tests run](https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/4222887432).

### Related issues, RFCs



**Issue number:** #1309 

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist


### Breaking change checklist


---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
